### ISSUE TITLE
Why is ClipRect the io.DisplaySize?

### ISSUE THREAD

**Issue by Mellnik** [ocornut/imgui#6280]
**Version/Branch of Dear ImGui:**

Version: 1.89.4
Branch: master

**Back-end/Renderer/Compiler/OS**

Back-ends: custom
Compiler: VS2022
Operating System: Windows 10

**My Issue/Question:**

One of the ImDrawCmd.ClipRect is the same size as io.DisplaySize. See the screenshot below.
The first drawcmd is the top header of the first demo window. The ClipRect is the same size as io.DisplaySize.

Why is that?

In my custom backend I tell the engine where stuff is drawn so that the mouse input system can detect what a user wants to interact with using the mouse. This means there is other GUI stuff drawn besides Dear ImGui.

I guess this on purpose but how would I get only the rectangles of the stuff that Dear ImGui draws?

Thank you!

**Screenshots/Video**

![cliprect](https://user-images.githubusercontent.com/2716498/228485619-8424d3d0-e788-4dc1-bfce-363187819573.PNG)

**Comment by Mellnik** [comment]
Thank you for your reply.

I've read this part and I totally understand how it is done in your described case.

But please imagine the following scenario. I am working with the [Unreal Engine implementation by segross](https://github.com/segross/UnrealImGui).

In UE there is a widget system. Each UI is rendered in a widget instance which has a specific size/hittest for the mouse.
When the user clicks on anything on the screen the Unreal Engine does a hittest on what widget is on the top.
The widget that was hit receives focus and **therefore keyboard and mouse input on the widget instance.** 

Now when I render Dear ImGui on a full screen widget it is impossible that any widget that is below ImGui ever receives input.

Does this make sense to you? Maybe I am overseeing something....

This is the state of the implementation by segross. Unfortunately he also seems to no longer work on it so I cannot cannot contact him.

Anyway thank you for your work and time.

**Comment by ocornut** [comment]
Supposedly if you are trying to hit-test for window geometry (which is what you were trying to do?) it seems you would have a way to direct that input to the underlying contents. Therefore I don't understand the relation between the initial question and your second post.

Or do you want you want that geometry in order to be able to submit the rectangles to some Unreal hit-testing system so it can decide itself ?

**Comment by Mellnik** [comment]
> Or do you want you want that geometry in order to be able to submit the rectangles to some Unreal hit-testing system so it can decide itself ?

Yes exactly! I want to submit the rectangles to Unreal's hittesting system so that it can decide.

**Comment by ocornut** [comment]
(To answer the question as asked: the first ImDrawCmd of most windows has ClipRect = (0..io.DisplaySize) because well, we want that window decoration to appear and not be clipped... Most often when the window has contents and for the benefit of scrolling, the second ImDrawCmd emitted use a ClipRect that matches the inner client area of the window. But I don't think this has anything to do with your problem)

**Comment by ocornut** [comment]
> Yes exactly! I want to submit the rectangles to Unreal's hittesting system so that it can decide.

Honestly this seem like asking for trouble, but ok.

I think you might be able to use `FindHoveredWindow()` in imgui.cpp as a very rough reference.

Basically you'd want to:
- Iterate every `g.WindowsFocusOrder[]` (this will be faster than `g.Windows[]` as you only need root windows).
- Ignore those where `!window->Active || window->Hidden`.
- Pull the `window->OuterRectClipped` rectangle.
- Add padding as done in `FindHoveredWindow()`.
- Rare and used by docking only: if `window->HitTestHoleSize != 0)` the window may have a hit-test rectangle hole. In this case instead of using the outer rectangle, extrude `HitTestHoleXXX` away from it and submit 8 smaller rectangles. As a convenience you can refer to drawing code `RenderRectFilledWithHole()` of how to extract those 8 smaller rectangles out of 1 big + 1 hole.

I think you can write a single function that takes that info and output e.g. `ImVector<ImRect>` which you will submit to Unreal. You can use `GetForegroundDrawList()->AddRectFilled(....., IM_COL32(0,255,0,50))` to visualize the output of your function, and if it works it'd be nice to post it here.

**Comment by Mellnik** [comment]
Thanks a lot! I am going to look into that and work on the implementation. Will post once everything is working as expected.

**Comment by Mellnik** [comment]
Alright after investigating more and after your answers I found a better solution. Using the custom hitrects will probably ask for trouble as you said.
But your explanation on getting the rects is useful nevertheless for other people as well.
Thanks again and I will close this issue.

**Comment by ocornut** [comment]
May I ask which better solution you did find?

**Comment by Mellnik** [comment]
Sure. I found a way to make it work with the standard behavior that you linked above. (https://github.com/ocornut/imgui/blob/master/docs/FAQ.md#q-how-can-i-tell-whether-to-dispatch-mousekeyboard-to-dear-imgui-or-my-application)

1. Getting mouse input outside of the Unreal widget and passing it to ImGui. = Always passing mouse move inputs to ImGui even if the widget has no focus.
2. After `ImGui::Render` and `ImGui::GetDrawData`` check for `io.WantCaptureMouse`. If it is `false` then I simply remove the widget from Unreal's hittesting system. This allows widgets below the ImGui fullscreen widget to receive input.
